pandas/tests/groupby/test_groupby.py

```diff
@@ -944,16 +947,28 @@ def f_no_copy(x):
     assert_series_equal(grpby_copy, grpby_no_copy)
 
 
-def test_no_mutate_but_looks_like():
+def test_no_mutate_but_looks_like(as_index):
 
     # GH 8467
     # first show's mutation indicator
     # second does not, but should yield the same results
     df = DataFrame({"key": [1, 1, 1, 2, 2, 2, 3, 3, 3], "value": range(9)})
 
-    result1 = df.groupby("key", group_keys=True).apply(lambda x: x[:].key)
-    result2 = df.groupby("key", group_keys=True).apply(lambda x: x.key)
-    assert_series_equal(result1, result2)
+    def run_test(df, as_index):
```

Comment:
do not nest functions like this; best to split this into multiple tests and/or parameterize